### ISSUE TITLE
podman seems to "remember" what was the latest --arch flag and implicitly reuse it.

### ISSUE THREAD

**Issue by Romain-Geissler-1A** [containers/podman#12682]
/kind bug

**Description**

Podman seems to "remember" what was the latest --arch flag, and use it implicitly if it was not provided on the command line.

**Steps to reproduce the issue:**

1. On a RHEL 8 where podman and qemu-user-static is install, first run a container with an explicit arch, for example amd64:
```
rgeissler@NCEOBERHEL80009.rnd.amadeus.net[dev]:~$ podman run --arch=amd64 --rm fedora uname -m
Resolved "fedora" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
Trying to pull registry.fedoraproject.org/fedora:latest...
Getting image source signatures
Copying blob 4545346f2a49 [--------------------------------------] 0.0b / 0.0b
Copying config 3059bef432 done
Writing manifest to image destination
Storing signatures
x86_64
```

2. Now run the same command, without explicit --arch: it still uses x86_64, which in my case is the native arch, it is expected:
```
rgeissler@NCEOBERHEL80009.rnd.amadeus.net[dev]:~$ podman run --rm fedora uname -m
x86_64
```

3. Now run explicitly an aarch64 container, emulated with qemu:
```
rgeissler@NCEOBERHEL80009.rnd.amadeus.net[dev]:~$ podman run --arch=arm64 --rm fedora uname -m
Resolved "fedora" as an alias (/etc/containers/registries.conf.d/000-shortnames.conf)
Trying to pull registry.fedoraproject.org/fedora:latest...
Getting image source signatures
Copying blob 73333db80aa9 [--------------------------------------] 0.0b / 0.0b
Copying config b74713569c done
Writing manifest to image destination
Storing signatures
aarch64
```

4. Finally run a container again, without explicit --arch: it runs as aarch64, while it should be x86_64:
```
rgeissler@NCEOBERHEL80009.rnd.amadeus.net[dev]:~$ podman run --rm fedora uname -m
aarch64
```

**Describe the results you received:**

In the 4th step, container is run as aarch64, while it should be x86_64.

**Describe the results you expected:**

In the 4th step, container is run as aarch64, while it should be x86_64.

**Additional information you deem important (e.g. issue happens only occasionally):**

Since Red Hat doesn't provide official qemu-user-static packages, we actually did install the fedora one on RHEL 8. I know it's not covered by the Red Hat support policy, but most likely running all this on a fedora machine would yield similar results, I doubt it's related to RHEL 8, more to podman.

**Output of `podman version`:**

```
rgeissler@NCEOBERHEL80009.rnd.amadeus.net[dev]:~$ podman version
Version:      3.3.1
API Version:  3.3.1
Go Version:   go1.16.7
Built:        Tue Sep 21 08:41:42 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
rgeissler@NCEOBERHEL80009.rnd.amadeus.net[dev]:~$ podman info --debug
host:
  arch: amd64
  buildahVersion: 1.22.3
  cgroupControllers: []
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.29-1.module+el8.5.0+12582+56d94c81.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.29, commit: 0f5bee61b18d4581668e5bf18b910cda3cff5081'
  cpus: 12
  distribution:
    distribution: '"rhel"'
    version: "8.5"
  eventLogger: file
  hostname: NCEOBERHEL80009
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 10011
      size: 1
    - container_id: 1
      host_id: 2548576
      size: 500000
    uidmap:
    - container_id: 0
      host_id: 50515
      size: 1
    - container_id: 1
      host_id: 2548576
      size: 500000
  kernel: 4.18.0-348.el8.x86_64
  linkmode: dynamic
  memFree: 20370321408
  memTotal: 25256333312
  ociRuntime:
    name: runc   
    package: runc-1.0.2-1.module+el8.5.0+12582+56d94c81.x86_64
    path: /usr/bin/runc
    version: |-  
      runc version 1.0.2
      spec: 1.0.2-dev
      go: go1.16.7
      libseccomp: 2.5.1
  os: linux
  remoteSocket:  
    exists: true 
    path: /run/user/50515/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_NET_RAW,CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:   
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.8-1.module+el8.5.0+12582+56d94c81.x86_64
    version: |-  
      slirp4netns version 1.1.8
      commit: d361001f495417b880f20329121e3aa431a8f90f
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.1
  swapFree: 1073737728
  swapTotal: 1073737728
  uptime: 166h 53m 44.15s (Approximately 6.92 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /srv/data/home/rgeissler/.config/containers/storage.conf
  containerStore:
    number: 1
    paused: 0
    running: 1   
    stopped: 0   
  graphDriverName: overlay
  graphOptions:  
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.7.1-1.module+el8.5.0+12582+56d94c81.x86_64
      Version: |-
        fusermount3 version: 3.2.1
        fuse-overlayfs: version 1.7.1
        FUSE library version 3.2.1
        using FUSE kernel interface version 7.26
  graphRoot: /srv/data/home/rgeissler/.local/share/containers/storage
  graphStatus:   
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 8
  runRoot: /run/user/50515/containers
  volumePath: /srv/data/home/rgeissler/.local/share/containers/storage/volumes
version:
  APIVersion: 3.3.1
  Built: 1632213702
  BuiltTime: Tue Sep 21 08:41:42 2021
  GitCommit: ""  
  GoVersion: go1.16.7
  OsArch: linux/amd64
  Version: 3.3.1

```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
rgeissler@NCEOBERHEL80009.rnd.amadeus.net[dev]:~$ rpm -q podman
podman-3.3.1-9.module+el8.5.0+12697+018f24d7.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

No

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical x86_64 RHEL 8 machine, with qemu-user-static from fedora install on it.

**Comment by vrothberg** [comment]
Thanks for taking the time to open the issue.

Yes, `podman run $IMAGE` will run `$IMAGE` independent of the architecture. While this seems strange at first glance, it is an unfortunate necessity since there are images which specify the wrong architecture (see https://github.com/containers/podman/issues/10682).  That means that once an image has been pulled, Podman will not apply any architecture checks when looking it up again. To resolve your issue and pull down the "correct" image, you can either run `podman pull $IMAGE` or `podman run --always $IMAGE` or `podman run --arch $ARCH $IMAGE`. Note that when `$ARCH` is specified, Podman will also pessimistically pull the image down.

Docker does not have this issue as Docker does not support running multi-arch containers.

**Comment by Romain-Geissler-1A** [comment]
Mmmh, this is indeed very counter intuitive. Given that on my account, I am run many scripts from many projects that eventually start a container, and since I tend to use fedora by default in a lot of my projects, it means that to be shielded by this I shall add --arch $(uname -m) in pretty much all my "podman/docker run" command in all my projects, I don't find this super user friendly.

However I just checked (since you said Docker can't run multi arch containers), actually:
 - Docker can run multi-arch containers
 - Docker has *EXACTLY* the same behavior than podman:

```
ubuntu@olaf:~> docker run -t -i --rm --platform aarch64 fedora uname -a
Unable to find image 'fedora:latest' locally
latest: Pulling from library/fedora
3f28ea9d8c33: Pull complete
Digest: sha256:40ba585f0e25c096a08c30ab2f70ef3820b8ea5a4bdd16da0edbfc0a6952fa57
Status: Downloaded newer image for fedora:latest
Linux 6f0b4e3d5ead 5.4.0-91-generic #102-Ubuntu SMP Fri Nov 5 16:31:28 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
ubuntu@olaf:~> docker run -t -i --rm fedora uname -a
WARNING: The requested image's platform (linux/arm64) does not match the detected host platform (linux/amd64) and no specific platform was requested
Linux ea3a85ed4482 5.4.0-91-generic #102-Ubuntu SMP Fri Nov 5 16:31:28 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
```

Ok, l accept the answer, but honestly I find this behavior strange (though I understand you did code it because you liked it, but to fix wrong multi-arch image).

**Comment by vrothberg** [comment]
Thanks for your understanding, @Romain-Geissler-1A!  Also thanks for correcting, I was entirely unaware that Docker is supporting it as well.

Would a WARNING as emitted by Docker be more user friendly in your eyes?  I totally sympathize that it's not ideal for you use case but I think that our hands are tied given that so many images don't get the architecture right.

**Comment by Romain-Geissler-1A** [comment]
Yes I consider that the warning that docker emits is welcome in understanding a bit more what's wrong. I would personally add in this warning message something mentioning that either you shall explicitly provide an arch (what docker reads already), or somehow re-pull/change the pull policy to always pull, since both can fix the issue.

And from now on internally in Amadeus, since we are about to migrate partially our workloads to arm, I will do some internal advocacy so that people *ALWAYS* define an explicit --arch from now one in all their podman/docker command (and somehow I need to implement https://github.com/containers/podman/issues/12680 so that people can easily use "--arch $(uname -m)" that works with Docker (docker understands only --platform and not --arch, but "--platform $(uname -m)" works).

**Comment by vrothberg** [comment]
@Romain-Geissler-1A, maybe I can give you more. How about a new option in `/etc/containers/containers.conf` that would allow you enforce the arch/platform checks when looking up local images? Probably something we can do some plumbing for on the CLI as well.

@rhatdan PTAL ... mutli-arch is hard and I think we should continue doing what we're doing BUT an option can be a nice compromise if users know what they're doining.

**Comment by vrothberg** [comment]
@Romain-Geissler-1A, I reopened the issue and turned it into a feature request.

**Comment by Romain-Geissler-1A** [comment]
Just throwing here some random ideas. Do you have a way to store in the local image cache what was the full collection of variants offered by a given image name ? Typically, for an image that has only a single variant, even different from the host one (for example an arm image I have just built locally), podman/docker shall always attempt to run it with that unique arch if no --arch flag was provided on the command line. On the other side, an image that you know has several variants, among which one of them matches the host, should result in podman/docker trying to run that host arch if no --arch was provided. And finally in the case of an image with several variants, but none of them matching the host, podman/docker can stick the the existing behavior: run the variant that was latest pulled in the image cache.

But all this relies on the fact that you get informations about the manifest.list for images you pulled locally already, and I am not sure this is the case with your current storage mechanism.

**Comment by vrothberg** [comment]
> Just throwing here some random ideas. Do you have a way to store in the local image cache what was the full collection of variants offered by a given image name ?

The information if a given image has multiple variants is not generally available since images can also be pulled directly by digest which will pull the image directly without looking at the manifest list (Docker) or image index (OCI).

Assuming we had such information, I would still be extremely careful since images may change on the source registry.

I think the core of the problem is that `podman run $IMAGE` _should_ ideally perform architecture/platform checks but those are unreliable given so many images in the wild get it wrong.

> @Romain-Geissler-1A, maybe I can give you more. How about a new option in /etc/containers/containers.conf that would allow you enforce the arch/platform checks when looking up local images? Probably something we can do some plumbing for on the CLI as well.

Would that help your organization?  Enforcing platform checks would work assuming all your images get things right.

**Comment by Romain-Geissler-1A** [comment]
> Would that help your organization? Enforcing platform checks would work assuming all your images get things right.

If such a strict option existed, yes I would definitely reach the people in charge of the developer VM management to customize our podman installation with this stricter option enabled in /etc/containers/containers.conf. In case of podman remote, I am not sure if the client side /etc/containers/containers.conf would be read, or the server side /etc/containers/containers.conf. However, since I see we are doing more and more podman in podman/docker scenarios (since we are basically still stuck with only docker daemon in our CI for now, at whole company level), it means many people will install podman themselves, and so would be unlikely to enable this stricter option.

I don't know what is the amount of wrong image in the wild, but usually when I have to make such a decision for users of my software, I try to favor the legit users rather than the ones running non valid cases (like the ones having multi arch image with wrong arch). So, if the impact is acceptable (which I can't judge myself), I would rather make the strict behavior of podman the default one, and advocate that the ones having invalid images to use an opt-in option to lower the checks run by podman.

**Comment by rhatdan** [comment]
I am fine with adding a strict check, I even think it is a good idea.  Maybe at some point in the distant future we could turn it on by default.

**Comment by rhatdan** [comment]
@Romain-Geissler-1A Interested in opening a PR?

**Comment by paravoid** [comment]
I'm also bitten by this. The summary is that a `podman run --arch foo` run is effectively "sticky" and overrides the architecture option of any future `podman run` (without an `--arch` argument), which is very counter-intuitive.

Something that perhaps isn't clear in the report so far, is that in addition to that, the output of `podman image ls` becomes quite confusing after a `podman run --arch`:

```
host:~$ uname -m
x86_64

host:~$ podman image ls
REPOSITORY                     TAG         IMAGE ID      CREATED       SIZE
docker.io/library/debian       bullseye    04fbdaf87a6a  37 hours ago  129 MB

host:~$ podman run -it --rm debian:bullseye
root@a98b114a634d:/# uname -m
x86_64
root@a68dddb73564:/# 
exit

host:~$ podman run --arch arm64 -it --rm debian:bullseye
Resolved "debian" as an alias (/etc/containers/registries.conf.d/shortnames.conf)
Trying to pull docker.io/library/debian:bullseye...
Getting image source signatures
Copying blob 39ab78bc09e7 [--------------------------------------] 0.0b / 0.0b
Copying config 0371e3756e done  
Writing manifest to image destination
Storing signatures
root@7a611a762531:/# uname -m
aarch64
root@7a611a762531:/# 
exit

host:~$ podman image ls
REPOSITORY                     TAG         IMAGE ID      CREATED       SIZE
docker.io/library/debian       bullseye    0371e3756e46  37 hours ago  123 MB
<none>                         <none>      04fbdaf87a6a  37 hours ago  129 MB
```

Note how debian:bullseye now resolves to 0371e3756e46 (the arm64 variant), and 04fbdaf87a6a (the amd64 variant of debian:bullseye) now shows up as "<none>", whereas before the last run, it showed up correctly as debian:bullseye

 From a UX perspective, I think the most intuitive would be to show both as debian:bullseye, and have another column of "ARCH" clarifying the selected architecture (perhaps hidden behind an option, or even better, only showing up if there are multiple architectures for a given image, or if the architecture is foreign to the native one).

**Comment by rhatdan** [comment]
@vrothberg Did we make any progress on this?

**Comment by vrothberg** [comment]
> @vrothberg Did we make any progress on this?

No.  I think we need to take a step back and discuss how we want to handle multi-arch as whole.  Once we have a vision, we can check what needs to done.

**Comment by rhatdan** [comment]
@vrothberg what should we do with this one?

**Comment by vrothberg** [comment]
https://github.com/containers/podman/issues/14271#issuecomment-1129747114 asks for the same checks but when _pulling_ an image.

**Comment by vrothberg** [comment]
I am assigning the issue to me to easier find it again.  I did not start working on it yet, so if others want to tackle it, please free to do so and drop a comment here.

**Comment by fishy** [comment]
@vrothberg I verified that this is indeed fixed in podman 4.3.1 (using `--platform="linux/amd64"` no longer breaks the local cache, as I described in #14197). But I also noticed that adding `--platform="linux/amd64"` makes podman to run significantly slower than without `--platform` arg (my local machine is already linux/amd64). is that expected?

**Comment by rhatdan** [comment]
No it should not.

**Comment by fishy** [comment]
> No it should not.

Thanks. Filed https://github.com/containers/podman/issues/17063
